### PR TITLE
HostName and getHostNameInfo

### PR DESCRIPTION
**Pull request by NicolasDP**
add function to retrieve addresses from a given hostname

```haskell
addrs <- (addresses <$> getAddressInfo "github.com") `catch` \_ -> return [] :: IO [IPv4]
case addrs of
    [] -> error "no knonw address"
    i:is -> putStrLn $ "IPv4 " <> toString i
```

- [ ] missing test
- [x] move the c_get_h_errno in the Foundation's binding
- [x] use `safe` instead of `unsafe` for the gethostbyname2 FFI

### CODE REVIEW

File: Foundation/Network/HostName.hsc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+#include <winsock2.h>
+#else
+#include "netinet/in.h"
+#include "netdb.h"
+#endif
+
+-- | HostName
+--
+newtype HostName = HostName { toString :: String }
+  deriving (Show, Eq, Ord, Typeable, Hashable)
```
Comment (vincenthz) [inline Foundation/Network/HostName.hsc]: I think it make sense to not derive Show automatically, as this is a primitive type. hostname should be displayed as if it was a String itself

File: Foundation/Network/HostName.hsc [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+-- re-entrant function `gethostbyname2`. This function is using a `MVar ()` to
+-- avoid a race condition and should be safe to use.
+--
+getHostNameInfo :: (Show address_type, Eq address_type, Address address_type)
+                => HostName
+                -> IO (HostNameInfo address_type)
+getHostNameInfo = getHostNameInfo_ Proxy
+
+globalMutex :: MVar ()
+globalMutex = unsafePerformIO (newMVar ())
```
Comment (vincenthz) [inline Foundation/Network/HostName.hsc]: probably missing a NOINLINE pragma

File: Foundation/Network/HostName.hsc [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        return $ HostNameInfo (HostName on) as addrs
+  where
+    loop f = do
+        ptr <- f
+        if ptr /= nullPtr
+            then return ptr
+            else do
+                err <- c_get_h_errno
+                case err of
+                    _ | err == (#const HOST_NOT_FOUND) -> throwIO $ HostNotFound h
```
Comment (vincenthz) [inline Foundation/Network/HostName.hsc]: errno are available from `Foundation.System.Bindings.Posix`

Comment (NicolasDP) [inline Foundation/Network/HostName.hsc]: this is not errno

Comment (vincenthz) [inline Foundation/Network/HostName.hsc]: yep, didn't really noticed the h. it really need a better bindings, and moving to System.Bindings too.

File: Foundation/Network/HostName.hsc (hunk 2 above)
Comment (NicolasDP) [inline Foundation/Network/HostName.hsc]: thanks

File: Foundation/Network/HostName.hsc (hunk 1 above)
Comment (NicolasDP) [inline Foundation/Network/HostName.hsc]: Good idea

File: Foundation/Network/HostName.hsc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+             -> IO [address_type]
+getAddresses p ty len ptr = do
+    x <- peek ptr
+    if x == any
+        then return []
+        else (:) x <$> getAddresses p ty len (ptr `plusPtr` 1)
+
+foreign import ccall unsafe "gethostbyname2"
+    c_gethostbyname2 :: CString -> CInt -> IO (Ptr Word8)
+foreign import ccall unsafe "get_h_errno"
```
Comment (NicolasDP) [inline Foundation/Network/HostName.hsc]: not really, it is not the same error code (`errno /= h_errno`). I will move the code in the Foundation's Sys binding.

File: Foundation/Network/Address.hsc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+#ifdef mingw32_HOST_OS
+#include <winsock2.h>
+#else
+#include "netinet/in.h"
+#endif
+
+-- | Network Address class
+--
+class StorableFixed addr => Address addr where
```
Comment (vincenthz) [inline Foundation/Network/Address.hsc]: I don't think `Address` is useful

File: Foundation/Network/Address.hsc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    -- | retrieve associated Network Family code for this type of address
+    familyCode :: proxy addr -> CInt
```
Comment (vincenthz) [inline Foundation/Network/Address.hsc]: nothing should expose a family code directly

File: Foundation/System/Bindings/Network.hsc [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import Foreign.C.Types
+
+#ifdef mingw32_HOST_OS
+# include <winsock2.h>
+#else
+# include "netinet/in.h"
+# include "netdb.h"
+#endif
+
+herr_HostNotFound
```
Comment (vincenthz) [inline Foundation/System/Bindings/Network.hsc]: this doesn't follow the sys<X><FCT> naming from other modules.

Comment (vincenthz) [inline Foundation/System/Bindings/Network.hsc]: or `sys<X>_<DEFINE>`

Comment (NicolasDP) [inline Foundation/System/Bindings/Network.hsc]: I see where you come from but this is not really related to `errno`. Would rather like to avoid confusion, especially for something restricted to hostname and that would certainly not have any other usage in other modules ?

File: cbits/foundation_network.c [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#include "foundation_system.h"
+
+#if defined(FOUNDATION_SYSTEM_WINDOWS)
+# include <winsock2.h>
+#else
+# include "netdb.h"
+#endif
+
+int get_h_errno(void)
```
Comment (vincenthz) [inline cbits/foundation_network.c]: if it's exposed it ought to be called `foundation_`.
also if WSAGetLastError, bindings to it directly (`sysWindowsSAGetLastError`) and preventing a needless export on windows sounds useful.

File: foundation.cabal [hunk 9]
```diff
@@ -110,6 +112,7 @@ Library
                      Foundation.Collection.Foldable
                      Foundation.Collection.Mutable
                      Foundation.Collection.Zippable
+                     Foundation.Collection.Mapable
```
Comment (vincenthz) [inline foundation.cabal]: Mappable

**Comment by NicolasDP** [comment]
The feature is working and should be safe to use for now. However, there will certainly be a change soon as I will explore other options with `getaddrinfo`.
